packages/devextreme/js/core/utils/object.js

```diff
@@ -39,16 +39,29 @@ const orderEach = function(map, func) {
     }
 };
 
-const assignValueToProperty = function(target, property, value, assignByReference) {
+const assignValueToProperty = function(target, property, value, extendComplexObject, assignByReference, shouldCopyUndefined) {
     if(!assignByReference && variableWrapper.isWrapped(target[property])) {
         variableWrapper.assign(target[property], value);
+    } else if(!assignByReference && Array.isArray(value)) {
+        target[property] = value.map(item => {
+            let itemTarget = item;
+
+            if(Array.isArray(item)) {
+                itemTarget = [];
+            }
+            if(isObject(item)) {
+                itemTarget = {};
+            }
+
+            return deepExtendArraySafe(itemTarget, item, extendComplexObject, assignByReference, shouldCopyUndefined);
```

Comment:
Now the ```deepExtendArraySafe``` recursive function uses the ```assignValueToProperty``` function, that uses the ```deepExtendArraySafe``` inside again

Let's make the design of these functions better and move out this new block to the ```deepExtendArraySafe```

One more point for it, the recursive call of the ```deepExtendArraySafe``` doesn't fit well with the ```assignValueToProperty``` function title